README

$Id$

The asdcplib library is a set of objects that offer
simplified access to files conforming to the sound and
picture track file formats developed by the SMPTE Working
Group DC28.20 (now TC 21DC).

Recently, support has also been added for SMPTE draft ST
2067-5 "IMF Essence Component", AKA "AS-02".  This code was
donated by Fraunhofer IIS.  It carries additional copyright
information which should be listed whenever you link the
AS-02 elements of the library.  Please look at the top of
the AS-02 files to see this copyright information.

AS-02 support is carried in separate object modules, so
unless you #include <AS_02.h> and link libas-02.so you are
still using plain old asdcp.

This work was originally funded by Digital Cinema Initiatives,
LLC (DCI). Subsequent efforts have been funded by Deluxe
Laboratories, Doremi Labs, CineCert LLC, Avica Technology
and others.

**The asdcplib project was originally housed on SourceForge.
The project has moved to http://www.cinecert.com/asdcplib/

The project formerly depended upon the mxflib project. Because
of its focus on covering the whole of the MXF specifications,
mxflib is considerably larger and more complex that what I
require for this application. For this reason I have created
a dedicated MXF implementation that is now part of this
library. Special thanks to Matt Beard and Oliver Morgan for
their great work and support.

Thanks also to the members of the SMPTE DC28.20 packaging
ad-hoc group and the members of the MXF Interop Initiative
for their encouragement and support. Special thanks to
Jim Whittlesey and Howard Lukk at DCI for proposing and
supporting this project.


Design Notes

This library is intended (but of course not limited) for
use by developers of commercial D-Cinema products (and now
IMF!).  It is designed to be easily integrated into a wide
variety of development environments.  Commercial users are
strongly urged to use static linking (at least where you use
this library) to prevent malicious in-field replacement of
critical system modules. This recommendation should be
considered wherever Open Source or Free software is being
used in conjunction with critical security parameters, such
as cryptographic keys.

The author strives mightily to provide an API that is completely
independent of operating system and other library dependencies,
and which allows selective replacement of some modules for
local needs.  Specifically, the essence parsers and OpenSSL
crypto functions can be replaced by linking to alternative
implementations of the ASDCP:: objects which provide those
services.

AS_DCP.h contains the entire AS-DCP API.  You do not need to
read any of the other files, except maybe asdcp-test.cpp which
contains detailed usage examples of each of the API's services.
The KM_* files may be of interest for general development
support, but may be ignored if all you want is simple AS-DCP
support.

Likewise, draft 2067-5 "IMF Essence Component" (AS-02) support
is entirely contained in AS-02.h


Build Instructions

On more-or-less POSIX systems (OS X, Linux, and BSD), GNU make and 
autotools are required to build asdcplib. The same configure script 
will also build this package on Windows machines with Cygwin and
MinGW installed. For those Windows users who would prefer to build
this natively, an "nmake" build file and instructions can be found
in the win32 subdirectory.  

OpenSSL is also required, any recent version should be fine. See
http://www.openssl.org/ for more information and download instructions.

Optional support for writing Timed Text Track Files is supported by
either Xerces-C or Expat. See http://xerces.apache.org/xerces-c/ or
http://expat.sourceforge.net/ for source and build instructions.

To configure and build, type './configure' followed by 'make'. There
are several test targets on the POSIX side, but you need to assemble
a set of test files to use them.  AS-02 support can be enabled with 
 --enable-as-02.  

I have tested this build on win32, Linux, OpenBSD, and Darwin
platforms. Others may work as well.




Utilities

asdcp-test -  Writes, reads and verifies AS-DCP (MXF) track files.

asdcp-wrap - Writes AS-DCP (MXF) track files.

asdcp-unwrap - Extracts essence from AS-DCP (MXF) track files.

asdcp-info - Displays information about AS-DCP (MXF) track files.

asdcp-util - Calculates digests and generates random numbers and UUIDs.

as-02-wrap - Writes AS-02 Essence Component files.

as-02-unwrap - Extracts essence from AS-02 Essence Component files.

kmfilegen - Writes and verifies large files using a platform-
independent format. Use it to test issues related to large files.

kmuuidgen, kmrandgen - generate UUID values and random numbers.

wavesplit - Splits a WAVE file into two or more output files. Used
  to untangle incorrectly-paired DCDM sound files.

blackwave - Write a WAVE file of zeros.

pinkwave - Write a WAVE file of SMPTE ST 2095 pink noise.

j2c-test - Displays information about JP2K codestreams.

Experimental feature: Prototype for High Dynamic Range is a wrapper
for the IMF application that allows JPEG-2000 codestreams to be paired
with opaque blobs of metadata.  AS-02 support must be enabled to
build this feature, so --enable-as-02 must be enabled if
--enable-phdr is to be used.  The following executable programs will be
built:

phdr-wrap - Writes AS-02 PHDR Essence Component files.

phdr-unwrap - Extracts essence from AS-02 PHDR Essence Component files.


Documentation

The API documentation is mostly in AS_DCP.h. and AS_02.h  Read those
files for a detailed description of the library's capabilities. Read
asdcp-*.cpp and as-02-*.cpp files for library usage examples. The
command-line utilities all respond to -h.


Change History

2019-02-12 - 2.10.32 -- bug fixes
 o Fixed parsing of as-02-wrap -tfr (ACES) option value
 o Fixed ISXD bugs:
    o Incorrect UTF_8_Text_DataEssenceCoding UL
    o Incorrect ISXDDataEssenceDescriptor_NamespaceURI UL
    o Incorrect use of FrameWrappedISXDData UL
    o Incorrect use of FrameWrappedISXDContainer UL
 o Added wide-char support for filenames on Win32, allows UTF-8 characters in
   calls to Kumu::File*::Open*(). The macro KM_WIN32_UTF8 must be defined to
   enable this feature.


2018-10-15 - 2.10.31
 o NOTE: Use of MXFGCGenericEssenceMultipleMappings in AS-02 files is not
   defined by standards and will soon be withdrawn from this library. If you
   would like to experience this change now, define the macro
   ASDCP_GCMULTI_PATCH
 o Removed PIMD support (unused prototype)
 o Returned the DataEssenceCoding UL version byte to "5" in Atmos files
 o Fixed ColorPrimaries_ACES UL version byte
 o Added "FSKSync" audio channel label to the MCA label generator
 o SMPTE RDD 45 requires a minimum BER encoding length of five (5) bytes for
   Content Element packets. The Write_EKLV_Packet() API has been modified to
   allow the specification of this value. All existing wrapping modules have
   been changed to conform to the new API, and all continue to use four (4)
   as the minimum BER encoding length. Write_EKLV_Packet() will continue to
   generate BER encodings with a length greater than the specified minimum for
   payloads that are larger than values encoded using the specified minimum.


2018-09-14 - 2.9.30
 o Fixes wrong MajorVersion in Preface
 o Adds VideoLineMap also to RGBA essence descriptor
 o Adds VideoLineMap to RGBA and CDCI descriptor only when option -l
   is present (disputable - but distinguished value {0, 0} is not
   allowed in closed and complete partitions)
 o Fixes issue with IndexRate not set in AS-02 J2K files
 o Additional fixes to minor version number in AS-02 partition packs
 o CMake brought up to date
 o Fixed erroneous WAV RF64 header interpretation
 o ACES AS-02 ST 2065-5 wrapper
 o ACES AS-02 ST 2065-5 unwrapper
 o ST 2067-50 Subdescriptors support
 o ST 2067-50 Target Frame wrapping support
 o ST 2067-50 Target Frame unwrapping support
 o Command line tools as-02-wrap, as-02-unwrap and as-02-info extended to
   support ACES, App#5 subdescriptors and Target Frames
 o Beta support for ST2065-5 AS_02 ACES wrapping and unwrapping including
   ST 2067-50 Subdescriptors and Target Frame support
 o Please report ACES-related issues to wolfgang.ruppel@hs-rm.de


2018-08-08 - 2.8.29
 o Modified AS_02::ST2052_TextParser::OpenRead signature
 o Fixed partition version error in AS-02 interior body partitions
 o Changed StartTimecode to 0 (zero) in the Source Package
 o Remove use of BN_init in OpenSSL API
 o XML parser now works correctly with Xerces-c 3.x
 o Improved MCA sound field expression parser
 o Added support for IMF Numbered Source Channel labels in MCA
 o Added CLI switches for MCAAudioContentKind and MCAAudioElementKind
 o Re-assigned asdcp-wrap option '-g' to '-x', added new option '-g' 
 o Added TLM marker check to j2c-test
 o Added to kmrandgen options '-C', '-w', '-W'
 o Added support for SMPTE RDD 47 "ISXD Track File"
 o Added generic support for SMPTE RP 2057 "Text-Based Metadata Carriage in MXF"
 o Re-factored AS-02 frame-wrap index write to use a common implementation
 o Re-factored support for SMPTE ST 410 to use a common implementation (affects
     AS-DCP and AS-02 timed-text MXF (SMPTE 429-5)
 o Patched several ambiguous integer casts.
 o Added new essence type identifiers ESS_AS02_ISXD and ESS_AS02_ACES
 o Fixed a bug wherein the Generic Container data track clip was instead being
     written as a DM track
 o Fixed UL values DCDataDescriptor and ContainerConstraintSubDescriptor to have
     a byte 6 value of 0x53 (Tag Set) instead of 0x07 (coding sentinel)
 o Added UL values 468 - 511 to the dictionary (including added ACES UL values)
 o Added ApplicationSchemes and ConformsToSpecifications to the Preface set
 o Added default initializer to MXF::LineMapPair
 o Added sets DescriptiveFramework, DescriptiveObject, TextBasedDMFramework,
     TextBasedObject, GenericStreamTextBasedSet and ISXDDataEssenceDescriptor,
     ACESPictureSubDescriptor, TargetFrameSubDescriptor
 o Added ISXD support, options '-c', '-g', '-G', '-v', '-q', '-u' to as-02-wrap
 o Added ISXD support and option '-g' to as-02-unwrap
 o Cleaned up definition of UL WaveAudioDescriptor_PeakEnvelopeData (currently
     unused)
 o Moved ReferenceAudioAlignmentLevel from WaveAudioDescriptor to
     GenericSoundEssenceDescriptor


2016-12-10 - MDD bug fix 2.7.19
 o The previous version (2.7.18) contained incorrect MDD.cpp and MDD.h
   files which affect the PIMF and JP2KEssenceCompression labels.


2016-12-10 - bug fixes 2.7.18
 o Fixed *all* Partiton version numbers in AS-02 files (the previous attempt
   only affected the header partition)
 o Improved IMSC-1 profile detection. May not yet be perfect, experimentation
   encouraged!
 o Added IMF App2e UL values and new MXF data types ColorPrimary and
   ThreeColorPrimaries (contributed by Schleich)
 o Added picture descriptor elements MasteringDisplayPrimaries,
   MasteringDisplayWhitePointChromaticity, MasteringDisplayMaximumLuminance,
   and MasteringDisplayMinimumLuminance
 o Added -o and -O options to as-02-wrap 
 o Fixed as-02-wrap option -p when making 4:4:4 files
 o Simplified as-02-wrap syntax description


2016-12-01 - bug fixes 2.6.17
 o Fixed Partiton and Preface version numbers in AS-02 files
 o Replaced WIN32 directory scanner with dirent_win.h
 o The NamespaceURI property of AS-02 timed text files has been exposed in the API
   and via as-02-wrap -P. This behavior replaces previous bad behavior, but puts
   responsibility for selecting the correct value on the operator.
 o Exposed CreatePNGNameId and CreateFontNameId subroutines in AS_02::TimedText
 o Adjusted UUID generation based on PNG and font names to exclude paths by
   inserting a call to PathBasename at each Id generation site
 o Fixed AS-02 timed text index partition
 o Fixed contents of generic partiton EssenceContainers property in AS-DCP timed
   text files
 o Added PixelLayout to RGBAEssenceDescriptor (contributed by Schleich)
 o Cleared up MXFGCP1FrameWrappedPictureElement / MXFGCP1FrameWrappedPictureElement
   ambiguity. PHDR continues to use MXFGCP1FrameWrappedPictureElement.
   MXFGCI1FrameWrappedPictureElement is supported for interlace.
 o Added support for 192, 200 and 240 fps images, includes 96, 100 and 120 fpe stereo
 o Added missing MCA properties to MCALabelSubDescriptor (contributed by Ruppel)
 o New MXF data type: LineMapPair
 o Added default 0,0 VideoLineMap value when wrapping CDCI with as-02-wrap
 o Added VideoLineMap property to GenericPictureEssenceDescriptor
 o Added timed text unwrap support to as-02-unwrap (contributed by Ruppel)
 o Added prototype PIDM dynamic metadata wrapping, config with --enable-phdr
   (Not for production use, caveat emptor, not the droids, etc.)


2016-08-11 - bug fixes, 2.5.15

 o Modified as-02-wrap to pass WhiteRefLev, BlackRefLev and ColorRange
    (contributed by Elmer Musser / Rod Bogart)
 o Fixed a bug in the index reader that allowed an out-of-bounds vector index
   to be used.
 o Modified the IndexEntry parser to ignore additional bytes belonging to slice
   values.


2016-05-06 - bug fixes, 2.5.14

 o Release 2.5.12 caused two problems with Atoms and DTS-X wrapping that are
   addressed by this release. The UL values that were changed to correct the
   DCData (ST 429-14) implementation had two unintended side-effects:
     1. v2.5.12 cannot read old DCData (and thus DTS-X) track files
     2. v2.5.12 creates broken Atmos track files (incorrect ULs)
   The fix for this involved the following changes:
     1. ASDCP::DCData::(MXFReader|MXFWriter) are no longer base classes of
        ASDCP::ATMOS::(MXFReader|MXFWriter).
     2. New UL values have been added to carry the original Atoms UL values:
        a. DolbyAtmosDCDataWrappingFrame
        b. DolbyAtmosDCDataEssence
        c. DolbyAtmosDCDataDescriptor
     3. ASDCP::ATMOS::(MXFReader|MXFWriter) have been modified to use the
        DolbyAtmosDCData* UL values.
     4. ASDCP::DCData::MXFReader will parse files having either DCData* or
        DolbyAtmosDCData* UL values.
     5. ASDCP::DCData:: MXFWriter continues to use DCData* UL values from ST
        429-14
  o Some optional parameters of the MXF metadata packs were being erroneously
    initialized with a value of zero (0), such as GenericPictureEssenceDescriptor
    ::SignalStandard, GenericPictureEssenceDescriptor::SampledXOffset(0), etc.
  o Added higher edit rates (96, 100, 120) to the cinema TimedText_Parser

  o Defined bool ASDCP::UL::operator<(const UL& rhs) const, which is needed to
    carry out the version-insensitive lookup policy when searching for ULs in a
    std::map<> of UL entries.
  o New files:  src/info.in src/CMakeLists.txt
  o Removed filess:  AS_DCP_DCData_internal.h


2016-03-17 - bug fixes, 2.5.12
 o Changed default UNIX file permissions to "666"
 o Changed default UNIX file permissions to "777"
 o Added missing initializer for caltime::offset
 o Fixed UL values for Aux Data files (per ST 429-14)
 o Added support for default TTML fonts (ST2052_TextParser)
 o Added Solaris patch to KM_fileio (Thanks Mikey!)
 o Fixed the version byte in UL DataEssenceCoding
 o Fixed duplicate linkage error for symbol MIME2str
 o Big patch from Wolfgang Ruppel:
   o Corrected erroneous fixed-tag values in ActiveHeight, ActiveWidth,
     ActiveXOffset, ActiveYOffset UL definitions.
   o Provided a stub implementation for Kumu::DirScannerEx (A real
     implementation will follow soon.)
   o Added as-02-info program
   o Added Seek() method to PCM parsers
   o Added CMakeLists.txt


2015-11-10 - bug fixes, IMF text, pink noise, 2.5.11
 o I moved personal dev environment from older gcc to newer clang.  Many
   small changes were made to satisfy the new compiler:
   - Altered many printf format codes to use the correct format for the
     given integer type.
   - Parenthesized some expressions to clarify previously ambiguous
     expectation about precedence.

CHANGES

Change History

2022-04-28 2.12.2 -- bug fixes, IMF JPEG XS, no-OpenSSL build
 o Fixes to JPEG XS wrapping contributed.
 o correctly handle WAVE files that don't end with the "data" chunk.
 o Modified the KLV parser to return RESULT_ALLOC instead of RESULT_FAIL when an
   attempt is made to read a packet that is larger than the internal limit. The
   file handle is Seek()ed to the first byte following the huge packet, allowing
   the caller to get over it and continue parsing MXF.
 o Added a new UL matching operator MatchIgnorePlaceholder()
 O new command as-02-wrap-iab contributed.
 o as-02-wrap changes contributed:
    update -D documentation
    add -J option for writing J2CLayout
    add additional predefined RGBALayout values
    add support for COLOR.7
 o Set StaticTrack.TrackID incrementally from 2 instead of hardcoded.
 o Templatize h__AS02WriteClip, move functions definition to header. Note that
   this depends on filereader pluggable at runtime commit as the reader is called
   as a pointer here.
 o IFileReaderFactory API contributed.
 o Fix compilation without OpenSSL.
 o New command as-02-wrap-jxs contributed.
 o API functions contributed to AS_02_IAB.h: AddDmsGenericPartUtf8Text,
   ReadGenericStreamPartitionPayload.
 o Added JPEG XS items to EssenceType_t.
 o CMake patches contributed for WITHOUT_XML option and building static libs
 o Fixed "Argument parser fails when first MCA label track is empty"
 o Add SLVS language assignment functionality
 o Added a Clone() method to InterchangeObject and derivatives that allows making
   a descriptor copy that is not owned by any other strucure. The caller must
   arrange to delete the object pointer returned.
 o Replaced reference-to-a-pointer-to-the-dictionary idiom with simle pointer.
   Removes some edge cases where the wrong dictionary was being used.
 o Increase header buffer size when wrapping 428-7.
 o ST 429-5 now limits timed text resources to 4095; error log and failure return
   when exceeded.
 o Fixed reading of AS-02 files larger than 4 GB.

2020-10-21 2.10.38 -- bug fixes, IMF IAB
 o Add new subroutine to calculate ContainerDuration from IndexTableSegment.
 o Fixed issues with ContainerDuration validation before use.
 o Fixed conditional XML parser support.
 o Add API call to expose the MIC key from an encrypted MXF.
 o Add Dolby labels to MDD.h/MDD.cpp.
 o Fixes to out-of-bounds read conditions.
 o Add support for multi-segment index tables to AS_02.
 o The PRNG implementation has been modified to use the newly-added internal AES
   and SHA-1 instead of relying on OpenSSL (or LibreSSL.) This paves the way for
   building the library without OpenSSL when cinema encryption is not required.
 o Added SHA-1 builtin code for build w/o OpenSSL.
 o Added built-in AES support to reduce external dependency.
 o Fix large file support on Windows.
 o Support for IAB Track Files (SMPTE ST 2067-201) contributed.
 o No longer emit Timecode Tracks when writing AS-02 files.
 o Fixed "ResourceID and AssetID not zeroed".

2020-02-10 - 2.10.35 -- HTJ2c support, minor extensions, bug fixes
 o Support for wrapping High-Throughput j2c contrinuted by Pierre Lemieux.
 o Added support for DisplayType, IntrinsicPictureResolution, ZPositionInUse
   properties in timed text track files.
 o Merged bug fixes and minor extensions from contibutors, see git log for detail.
 o Updated to ACES writer to add a ContainerConstraintsSubDescriptor (mandatory
   per ST 379-2)
 o Add MCA labels for Sign Language Video Stream.


2019-02-22 - 2.10.32 -- bug fixes
 o Fixed parsing of as-02-wrap -tfr (ACES) option value
 o Fixed ISXD bugs:
    o Incorrect UTF_8_Text_DataEssenceCoding UL
    o Incorrect ISXDDataEssenceDescriptor_NamespaceURI UL
    o Incorrect use of FrameWrappedISXDData UL
    o Incorrect use of FrameWrappedISXDContainer UL
 o Added wide-char support for filenames on Win32, allows UTF-8 characters in
   calls to Kumu::File*::Open*(). The macro KM_WIN32_UTF8 must be defined to
   enable this feature.
 o CMake fixed to extract version form configure.ac


2018-10-15 - 2.10.31
 o NOTE: Use of MXFGCGenericEssenceMultipleMappings in AS-02 files is not
   defined by standards and will soon be withdrawn from this library. If you
   would like to experience this change now, define the macro
   ASDCP_GCMULTI_PATCH
 o Removed PIMD support (unused prototype)
 o Returned the DataEssenceCoding UL version byte to "5" in Atmos files
 o Fixed ColorPrimaries_ACES UL version byte
 o Added "FSKSync" audio channel label to the MCA label generator
 o SMPTE RDD 45 requires a minimum BER encoding length of five (5) bytes for
   Content Element packets. The Write_EKLV_Packet() API has been modified to
   allow the specification of this value. All existing wrapping modules have
   been changed to conform to the new API, and all continue to use four (4)
   as the minimum BER encoding length. Write_EKLV_Packet() will continue to
   generate BER encodings with a length greater than the specified minimum for
   payloads that are larger than values encoded using the specified minimum.


2018-09-14 - 2.9.30
 o Fixes wrong MajorVersion in Preface
 o Adds VideoLineMap also to RGBA essence descriptor
 o Adds VideoLineMap to RGBA and CDCI descriptor only when option -l
   is present (disputable - but distinguished value {0, 0} is not
   allowed in closed and complete partitions)
 o Fixes issue with IndexRate not set in AS-02 J2K files
 o Additional fixes to minor version number in AS-02 partition packs
 o CMake brought up to date
 o Fixed erroneous WAV RF64 header interpretation
 o ACES AS-02 ST 2065-5 wrapper
 o ACES AS-02 ST 2065-5 unwrapper
 o ST 2067-50 Subdescriptors support
 o ST 2067-50 Target Frame wrapping support
 o ST 2067-50 Target Frame unwrapping support
 o Command line tools as-02-wrap, as-02-unwrap and as-02-info extended to
   support ACES, App#5 subdescriptors and Target Frames
 o Beta support for ST2065-5 AS_02 ACES wrapping and unwrapping including

src/JXS.h
/*
Copyright (c) 2005-2014, John Hurst,
Copyright (c) 2020, Thomas Richter Fraunhofer IIS,
Copyright (c) 2020, Christian Minuth Fraunhofer IIS,
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:
1. Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.
3. The name of the author may not be used to endorse or promote products
   derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
/* !\file    JXS.h
\version $Id$
\brief   JPEG XS constants and data structures

This is not a complete enumeration of all things JXS.There is just enough here to
support parsing picture metadata from a codestream header.
*/

#ifndef _JXS_H_
#define _JXS_H_

// AS_DCP.h is included only for it's base type definitions.
#include <KM_platform.h>
#include <KM_util.h>
#include <Metadata.h>
#include <assert.h>

namespace ASDCP
{
	namespace JXS
	{
		const ui32_t MaxComponents = 4; // ISO 21122-1 Annex A.2 up to 8 components
		const ui32_t MaxHorizontalLevels = 15;
		const ui32_t MaxVerticalLevels = 2;

#pragma pack(1)
		struct ImageComponent_t  // Essentially, a lookalike of the CDT marker, just with less bit-packing
		{
		  ui8_t Bc;  // Bitdepth (literal, not -1)
		  ui8_t Sx;
		  ui8_t Sy;  // Subsampling factors, horizontal and vertically. Bit-packed in the marker.
		};
#pragma pack()

                const byte_t Magic[] = { 0xff, 0x10, 0xff };

		bool lookup_ColorPrimaries(int value, ASDCP::UL& ul);
		bool lookup_TransferCharacteristic(int value, ASDCP::UL& ul);
		bool lookup_CodingEquations(int value, ASDCP::UL& ul);

		enum Marker_t
		{
			MRK_NIL = 0,
			MRK_SOC = 0xff10, // Start of codestream
			MRK_EOC = 0xff11, // End of codestream
			MRK_PIH = 0xff12, // Picture header
			MRK_CDT = 0xff13, // Component table
			MRK_WGT = 0xff14, // Weights table
			MRK_COM = 0xff15, // Extensions marker
			MRK_NLT = 0xff16, // Nonlinearity marker (21122-1 2nd)
			MRK_CWD = 0xff17, // Component-dependent decomposition marker (21122-1 2nd)
			MRK_CTS = 0xff18, // Colour transformation spec marker (21122-1 2nd)
			MRK_CRG = 0xff19, // Component registration marker (21122-1 2nd)
			MRK_SLH = 0xff20, // Slice header
			MRK_CAP = 0xff50  // Capabilities marker
		};

		const char* GetMarkerString(Marker_t m);

		//
		class Marker
		{
			KM_NO_COPY_CONSTRUCT(Marker);

		public:
			Marker_t m_Type;
			bool     m_IsSegment;
			ui32_t   m_DataSize;
			const byte_t* m_Data;

			Marker() : m_Type(MRK_NIL), m_IsSegment(false), m_DataSize(0), m_Data(0) {}
			~Marker() {}

			void Dump(FILE* stream = 0) const;
		};

		//
		ASDCP::Result_t GetNextMarker(const byte_t**, Marker&);

		// accessor objects for marker segments
		namespace Accessor
		{
			// capability marker
			class CAP
			{
				const byte_t* m_MarkerData;
				ui32_t m_DataSize;
				KM_NO_COPY_CONSTRUCT(CAP);
				CAP();

			public:
				CAP(const Marker& M)
				{
					assert(M.m_Type == MRK_CAP);
					m_MarkerData = M.m_Data;
					m_DataSize = M.m_DataSize -2;
				}

				~CAP() {}

				inline ui16_t Size()   const { return KM_i16_BE(*(ui16_t*)(m_MarkerData -2 )); }

				void Dump(FILE* stream = 0) const;
			};

			// Nonlinearity marker(21122 - 1 2nd)
			class NLT
			{
				const byte_t* m_MarkerData;
				KM_NO_COPY_CONSTRUCT(NLT);
				NLT();

			public:
				NLT(const Marker& M)
				{
					assert(M.m_Type == MRK_NLT);
					m_MarkerData = M.m_Data;
				}

				~NLT() {}

				inline ui16_t Size()   const { return KM_i16_BE(*(ui16_t*)(m_MarkerData - 2)); }

				void Dump(FILE* stream = 0) const;
			};

			// picture header
			class PIH
			{
				const byte_t* m_MarkerData;
				ui32_t        m_DataSize;
				KM_NO_COPY_CONSTRUCT(PIH);
				PIH();

			public:
				PIH(const Marker& M)
				{
					assert(M.m_Type == MRK_PIH);
					m_MarkerData = M.m_Data;
					m_DataSize = M.m_DataSize;
				}

				~PIH() {}

				inline ui16_t LpihSize()	const { return KM_i16_BE(*(ui16_t*)(m_MarkerData - 2)); }
				inline ui32_t LcodSize()	const { return KM_i32_BE(*(ui32_t*)(m_MarkerData)); }
				inline ui16_t Ppih()		const { return KM_i16_BE(*(ui16_t*)(m_MarkerData + 4)); }
				inline ui16_t Plev()		const { return KM_i16_BE(*(ui16_t*)(m_MarkerData + 6)); }
				inline ui16_t Wf()			const { return KM_i16_BE(*(ui16_t*)(m_MarkerData + 8)); }
				inline ui16_t Hf()			const { return KM_i16_BE(*(ui16_t*)(m_MarkerData + 10)); }
				inline ui16_t Cw()			const { return KM_i16_BE(*(ui16_t*)(m_MarkerData + 12)); }
				inline ui16_t Hsl()			const { return KM_i16_BE(*(ui16_t*)(m_MarkerData + 14)); }
				inline ui8_t  Nc()			const { return (*(ui8_t*)(m_MarkerData + 16)); }
				inline ui8_t  Ng()			const { return *(ui8_t*)(m_MarkerData + 17); }
				inline ui8_t  Ss()			const { return *(ui8_t*)(m_MarkerData + 18); }
				inline ui8_t  Cpih()		const { return (*(ui8_t*)(m_MarkerData + 21)) & 0x0f; }
				inline ui8_t  Nlx()			const { return (*(ui8_t*)(m_MarkerData + 22)) >> 4; }
				inline ui8_t  Nly()			const { return (*(ui8_t*)(m_MarkerData + 22)) & 0x0f; }

				void Dump(FILE* stream = 0) const;
			};

			class CDT
			{
				const byte_t* m_MarkerData;
				KM_NO_COPY_CONSTRUCT(CDT);
				CDT();

			public:
				CDT(const Marker& M)
				{
					assert(M.m_Type == MRK_CDT);
					m_MarkerData = M.m_Data;
				}

				~CDT() {}

				inline ui8_t  Bc(int i)          const { return *(m_MarkerData + (i << 1)); }
				inline ui8_t  Sx(int i)          const { return *(m_MarkerData + 1 + (i << 1)) >> 4; }
				inline ui8_t  Sy(int i)          const { return *(m_MarkerData + 1 + (i << 1)) & 0x0f; }

				void Dump(FILE* stream = 0) const;
			};

			class WGT
			{
				const byte_t* m_MarkerData;
				KM_NO_COPY_CONSTRUCT(WGT);
				WGT();

			public:
				WGT(const Marker& M)
				{
					assert(M.m_Type == MRK_WGT);
					m_MarkerData = M.m_Data;
				}

				~WGT() {}

				void Dump(FILE* stream = 0) const;
			};

			class COM
			{
				const byte_t* m_MarkerData;
				KM_NO_COPY_CONSTRUCT(COM);
				COM();

			public:
				COM(const Marker& M)
				{
					assert(M.m_Type == MRK_COM);
					m_MarkerData = M.m_Data;
				}

				~COM() {}

				void Dump(FILE* stream = 0) const;
			};
		}

	  //
	  class FrameBuffer : public ASDCP::FrameBuffer
	  {
	  public:
		  FrameBuffer() {}
		  FrameBuffer(ui32_t size) { Capacity(size); }
		  virtual ~FrameBuffer() {}

		  // Print debugging information to stream (stderr default)
		  void Dump(FILE* = 0, ui32_t dump_bytes = 0) const;
	  };

	  // An object which opens and reads a JPEG XS codestream file.  The file is expected
	  // to contain exactly one complete frame of picture essence as an unwrapped (raw)
	  // ISO/IEC 21122 codestream.
	  class CodestreamParser
	  {
		  class h__CodestreamParser;
		  mem_ptr<h__CodestreamParser> m_Parser;
		  ASDCP_NO_COPY_CONSTRUCT(CodestreamParser);

	  public:
		  CodestreamParser();
		  virtual ~CodestreamParser();

		  // Opens a file for reading, parses enough data to provide a complete
			  // set of stream metadata for the MXFWriter below.
		  // The frame buffer's PlaintextOffset parameter will be set to the first
		  // byte of the data segment. Set this value to zero if you want
		  // encrypted headers.
		  Result_t OpenReadFrame(const std::string& filename, FrameBuffer&) const;

		  // Fill the MXF descriptor objects with the values from the file's codestream.
		  // Returns RESULT_INIT if the file is not open.
		  Result_t FillPictureDescriptor(
                      ASDCP::MXF::GenericPictureEssenceDescriptor& picture_descriptor,
		      ASDCP::MXF::JPEGXSPictureSubDescriptor& jxs_subdescriptor) const;
	  };

	  // Parses the data in the frame buffer to fill in the picture descriptor. Copies
	  // the offset of the image data into start_of_data. Returns error if the parser fails.
	  Result_t ParseMetadataIntoDesc(const FrameBuffer&,
					 ASDCP::MXF::GenericPictureEssenceDescriptor& picture_descriptor,
					 ASDCP::MXF::JPEGXSPictureSubDescriptor& jxs_subdescriptor,
					 byte_t* start_of_data = 0);

	  // An object which reads a sequence of files containing JPEG XS pictures.
	  class SequenceParser
	  {
		  class h__SequenceParser;
		  mem_ptr<h__SequenceParser> m_Parser;
		  ASDCP_NO_COPY_CONSTRUCT(SequenceParser);

	  public:
		  SequenceParser();
		  virtual ~SequenceParser();

		  // Opens a directory for reading.  The directory is expected to contain one or
		  // more files, each containing the codestream for exactly one picture. The
		  // files must be named such that the frames are in temporal order when sorted
		  // alphabetically by filename. The parser will automatically parse enough data
		  // from the first file to provide a complete set of stream metadata for the
		  // MXFWriter below.  If the "pedantic" parameter is given and is true, the
		  // parser will check the metadata for each codestream and fail if a
		  // mismatch is detected.
		  Result_t OpenRead(const std::string& filename) const;

		  // Opens a file sequence for reading.  The sequence is expected to contain one or
		  // more filenames, each naming a file containing the codestream for exactly one
		  // picture. The parser will automatically parse enough data
		  // from the first file to provide a complete set of stream metadata for the
		  // MXFWriter below.  If the "pedantic" parameter is given and is true, the
		  // parser will check the metadata for each codestream and fail if a
		  // mismatch is detected.
		  Result_t OpenRead(const std::list<std::string>& file_list) const;

		  // Fill a PictureDescriptor struct with the values from the first file's codestream.
		  // Returns RESULT_INIT if the directory is not open.
		  Result_t FillPictureDescriptor(
                      ASDCP::MXF::GenericPictureEssenceDescriptor& picture_descriptor,
		      ASDCP::MXF::JPEGXSPictureSubDescriptor& jxs_subdescriptor) const;

		  // Rewind the directory to the beginning.
		  Result_t Reset() const;

		  // Reads the next sequential frame in the directory and places it in the
		  // frame buffer. Fails if the buffer is too small or the direcdtory
		  // contains no more files.
		  // The frame buffer's PlaintextOffset parameter will be set to the first
		  // byte of the data segment. Set this value to zero if you want
		  // encrypted headers.
		  Result_t ReadFrame(FrameBuffer&) const;
	  };

	} //namespace JXS
} // namespace ASDCP

#endif // _JXS_H_

//
// end JXS.h
//
